### DISCUSSION TITLE
Pycaret에 Custom KFold 추가하기

### DISCUSSION THREAD

**Discussion by Yuuraa** [bcaitech1/p4-dkt-freshtomato#81] [Update]
## 설명
마지막에 푼 문제만 validation으로 사용하는 KFold 입니다

## 코드
```python
# Custom Cross Validation - for pycaret
# validation 셋 안에서 user의 마지막 풀이 만을 가져왔다
from sklearn.model_selection import KFold

class LastSolveKFold:
    def __init__(self, last_indices, n_splits=10):
        self.last_indices = last_indices
        self.n_splits = n_splits

    def split(self, X, y, groups=None):
        for rx, tx in KFold(n_splits=self.n_splits).split(X,y):
            last_tx = np.array([idx for idx in tx if idx in self.last_indices])
            yield rx, last_tx


    def get_n_splits(self, X, y, groups=None):
        return self.n_splits
```
```python
last_indices = list(tr_val[tr_val['userID'] != tr_val['userID'].shift(-1)].index)
last_fold = LastSolveKFold(last_indices, n_splits=fold)
```

```python
def ensemble_automl(
                    tr_val,
                    holdout,
                    FEATS,
                    cat_features=[],
                    continuous_features=[],
                    seed=seed,
                    model_name_list=model_name_list,
                    ensemble_method = ensemble_method,
                    meta_model_for_stack = meta_model_for_stack,
                    ):

    # tr_val, holdout = datasets
    random.seed(seed)

    settings = setup(
                    data=tr_val[FEATS], 
                    target='answerCode', 
                    train_size=0.8, 
                    categorical_features=cat_features, 
                    numeric_features=continuous_features,
                    fold_strategy= last_fold, #< - 이 부분 추가!!
                    # log_experiment=False,   # compare_model 할 때에는 하이퍼파라미터가 어떻게 튜닝되었는지 볼 수 있어서 유용하다고 한다.
                    experiment_name=EXPERIMENT_NAME,
                    )

```

### Validation set의 갯수
10개 - 43, 42, 47, 47, 49, 48, 46, 44, 46, 47로 비교적 균등합니다

## 실행 결과 (재훈오빠 21개 features)
||Accuracy|AUC|Recall|Prec.|F1|Kappa|MCC|
|---|---|---|---|---|---|---|---|
|0|0.7353|0.8173|0.7692|0.8696|0.8163|0.3489|0.3575|
|1|0.7143|0.852|0.8125|0.65|0.7222|0.4355|0.447|
|2|0.6667|0.7846|0.85|0.68|0.7556|0.2515|0.2675|
|3|0.7692|0.7889|0.875|0.7778|0.8235|0.4935|0.5007|
|4|0.7105|0.7631|0.84|0.75|0.7925|0.3192|0.3249|
|5|0.7|0.7698|0.7826|0.72|0.75|0.3766|0.3787|
|6|0.8108|0.8939|0.8667|0.7222|0.7879|0.6197|0.6281|
|7|0.7368|0.8185|0.875|0.75|0.8077|0.3987|0.4108|
|8|0.7895|0.8145|0.8261|0.8261|0.8261|0.5594|0.5594|
|9|0.7059|0.8166|0.7059|0.7059|0.7059|0.4118|0.4118|
|Mean|0.7339|0.8119|0.8203|0.7452|0.7788|0.4215|0.4286|
|SD|0.042|0.0372|0.0516|0.0626|0.0408|0.1053|0.1036|

![image](https://user-images.githubusercontent.com/48202492/121013422-c4e06800-c7d3-11eb-8858-98e3aa41b738.png)

![image](https://user-images.githubusercontent.com/48202492/121013614-053fe600-c7d4-11eb-9517-e1deeab46736.png)

> CV 평균과 LB AUC 차이: 0.8119 - 0.8035 = 0. 0084
